tt_eager/tt_dnn/op_library/sdpa/kernels/dataflow/writer_decode_reducer_interleaved.cpp

```diff
@@ -0,0 +1,148 @@
+
+    constexpr bool is_dram = true;
+    constexpr uint32_t cb_out = tt::CB::c_out4;
+    constexpr uint32_t cb_intermed_out = tt::CB::c_out3;  // this cb holds the output intermediates from other worker cores
+    constexpr uint32_t cb_out_o = tt::CB::c_out0;
+    constexpr uint32_t cb_m_in = tt::CB::c_in6;
+    constexpr uint32_t cb_l_in = tt::CB::c_in7;
+
+    constexpr uint32_t tile_bytes = get_tile_size(cb_out);
+    constexpr uint32_t tile_bytes_intermed = get_tile_size(cb_intermed_out);
+    constexpr DataFormat data_format = get_dataformat(cb_out);
+
+    const InterleavedAddrGenFast<is_dram> out_writer = {
+        .bank_base_address = out_addr,
+        .page_size = tile_bytes,
+        .data_format = data_format
+    };
+
+    uint64_t intermed_l1_read_addr = get_noc_addr(get_read_ptr(cb_intermed_out));
+
+    volatile tt_l1_ptr uint32_t* in0_receiver_semaphore_addr_ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(in0_receiver_semaphore_addr);
+
+    constexpr uint32_t barrier_threshold = get_barrier_read_threshold<tile_bytes, num_cores>();
+    uint32_t barrier_count = 0;
+
+    constexpr uint32_t cb_scale_in = tt::CB::c_in4;
+    constexpr uint32_t cb_identity_scale_in = tt::CB::c_in5;
+
+    // generate and send scalar to compute
+    generate_bcast_unary_scalar(cb_scale_in, scale_val);
+    generate_reduce_scaler(cb_identity_scale_in, identity_scalar_packed);
+
+    // DPRINT << "[Writer Reducer] Pushed statistics to copmute" << ENDL();
+
+    if (k_chunk_end - k_chunk_start < k_num_chunks){
+        // This indicates that there are computes done by other workers. Needs to wait for them and send to reducer's compute
+        // Wait for compute to deliver output chunk, and write to compute again for reduction
+        // data in cb_intermed_out is arranged as [o,m,l,o,m,l,...] with size (out_chunk_tiles + 2*PNHt)*num_cores_to_wait
+        // wait on in0 semaphore value to become VALID (set by sender)
+        // DPRINT << "[Writer Reducer] Waiting for semaphore to be set from "<< num_cores_to_wait << " cores" << ENDL();
+        noc_semaphore_wait(in0_receiver_semaphore_addr_ptr, num_cores_to_wait);
+        // DPRINT << "[Writer Reducer] Received signal that semaphore has set" << ENDL();
+        // noc_semaphore_set(in0_receiver_semaphore_addr_ptr, 0);
+
+        // cb_wait_front(cb_intermed_out, num_tiles_to_wait);
+        constexpr uint32_t q_read_size = out_chunk_tiles*tile_bytes_intermed;
+        constexpr uint32_t ml_read_size = PNHt*tile_bytes_intermed;
+        // DPRINT << "[Writer Reducer] Received intermediate chunks from worker cores" << ENDL();
+        // DPRINT << "[Writer Reducer] Sending intermediate chunks to compute" << ENDL();
+        for(uint32_t block = 0; block < num_cores_per_batch; ++block) {
+
+            // DPRINT << "[Writer Reducer] Iteration " << block << ENDL();
+            cb_reserve_back(cb_out_o, out_chunk_tiles);
```

Comment:
Could this copying of local data from buffer to buffer lead to perf issues? How does perf look if you remove these local writes?
You could have the writer_worker write directly into the correct location of the reduce core's CB

Reply:
I think this simplifies the read logic in compute kernel so it is more bug prune. Due to there can be arbitrary number of worker cores, I have to create a single cb to hold the worker results. It would be hard to read with the offsets in the compute kernel as it is already big. I found it much simpler if compute just pops designated cbs for the worker input and statistics each loop. I don't think it will be a big perf issue because intermediate output and statistics are small. I will perf it to confirm

Reply:
I don't see any noticeable difference with/without noc_async read at this copying local data. So I think it is safe to leave it this way.